### ISSUE TITLE
edns-addr breaks caching

### ISSUE THREAD

**Issue by timkgh** [AdguardTeam/dnsproxy#329]
dnsproxy v0.49.1 with the following configuration:

```
cache: true
cache-size: 4194304
cache-min-ttl: 300
cache-max-ttl: 86400
cache-optimistic: false
edns: true
edns-addr: "x.x.x.0"
```
where `x.x.x.0` is an IP at my ISP

The caching doesn't seem to work. If I keep querying for `google.com` using `dig`, each query is slow (20ms) and the returned TTL is always 300 instead of decreasing for each query. This makes me think each query goes to the upstream instead of being served from the local cache.

If I disable `edns-addr` then queries are fast (0ms as reported by dig) and the returned TTL decreases with each query as expected.

**Comment by Mizzick** [comment]
@timkgh It would be helpful if you provide more details of your case?

**Comment by timkgh** [comment]
I tested some more and here's what I discovered using `edns-addr: "66.60.200.0"`

```
2023/05/11 07:08:37 3436332#69 [debug] dnsproxy: setting ecs: 66.60.200.0/24
...
2023/05/11 07:08:37 3436332#69 [debug] dnsproxy: cache: ecs option in response: 66.60.192.0/19
```
And it's never served from the cache because of the ecs subnet difference in the request and response.

If I use a random `edns-addr: "1.2.3.0"` then it works, it is served from the cache because the ecs in request and response match.

It appears that the dnsproxy cache matching needs to be done based on a wider subnet match not on an exact /24 match. Or use the key from the request rather than response.
